File: 1 sem/homework6/task3/hw6t3/stack.cpp
#include <stdio.h>
#include "stack.h"

struct StackElement
{
	int value;
	StackElement *next;
};

struct Stack
{
	StackElement *head;
};

Stack *createStack()
{
	Stack *result = new Stack;
	result->head = NULL;
	return result;
};

void push(Stack *stack, int value)
{
	StackElement *newHead = new StackElement;
	newHead->next = stack->head;
	newHead->value = value;
	stack->head = newHead;
}

char pop(Stack *stack)
{
	if (stack->head != NULL)
	{
		char result = stack->head->value;
		StackElement *temp = stack->head;
		stack->head = stack->head->next;
		delete temp;
		return result;
	}
	else
		return NULL;
}

void deleteElements(StackElement *head)
{
	if (head != NULL)
	{
		deleteElements(head->next);
		delete head;
	}
}

void deleteStack(Stack *stack)
{
	deleteElements(stack->head);
	delete stack;
}

File: 2 sem/homework7/task1/Stack/hw7t1stack/src/hw7t1stack/Stack.java
package hw7t1stack;

public class Stack<T> {
    
    public Stack() {
        head = null;
    }
    
    /**
     * Добавление элемента
     * @param value значение добавляемого элемента
     */	
    public void push(T value) {
	StackElement newHead = new StackElement(value, head);
	head = newHead;
    }
	
    /**
     * Удаление элемента
     * @return удаляемый элемент
     * @throws EmptyStackException бросает исключение, если стек пуст
     */
    public T pop() throws EmptyStackException {
        if (isEmpty()) {
	    throw new EmptyStackException();
	}
	else {
            T result = head.value;
	    head = head.next;
	    return result;
	}
    }
	
    /**
     * Проверка стека на пустоту
     * @return true - стек пуст, false - стек не пуст
     */
    public boolean isEmpty() {
	return head == null;
    }
	
    /**
     * 
     * Вершина стека
     * @return значение вершины стека
     * @throws EmptyStackException бросает исключение, если стек пуст
     */
    public T top() throws EmptyStackException {
        if (!isEmpty()) {
            return head.value;
        }
        else {
            throw new EmptyStackException();
        }	
    }
    
    private StackElement head;
    
    private class StackElement {
       
        public StackElement(T value, StackElement next) {
            this.value = value;
            this.next = next;
        }
       
        private T value;
       
        private StackElement next;
    }
}
